tests/test_mypy.py

```diff
@@ -0,0 +1,163 @@
+"""This test file is ignored by mypy since we're using pytest-mypy-testing to
+assert the mypy errors that should be raised.
+
+Despite some mypy errors, some fields will still attempt proper value conversions
+to the proper type. If the given value cannot be converted, a ValueError is raised.
+See ``tests/test_conversion.py`` for the conversion checks.
+
+Assertions in mypy are done via these line comments:
+    # N: <msg> - assert a mypy note message
+    # W: <msg> - assert a mypy warning message
+    # E: <msg> - assert a mypy error message
+    # R: <msg> - assert a mypy note message Revealed type is '<msg>'.
+
+NOTE:
+    Pytest's parametrization doesn't work here due to the dynamic nature of
+    variables in each line. For-loops also doesn't work well since the type
+    checker needs to narrow down on the type within the test function's scope.
+"""
+
+import pytest
+from web_poet import HttpResponse
+
+from zyte_common_items import (
+    Breadcrumb,
+    Image,
+    Link,
+    Product,
+    ProductFromList,
+    ProductList,
+    ProductVariant,
+)
+
+URL = "https://www.some-page/"
+RESPONSE = HttpResponse(URL, b"")
+
+
+@pytest.mark.mypy_testing
+def test_assignment_image():
+    img_1 = Image(url=RESPONSE.url)
+    img_1.url = RESPONSE.url  # E: Incompatible types in assignment (expression has type "ResponseUrl", variable has type "str")
```

Comment:
Hm, why does it raise a mypy error? Isn't the idea to support conversion on assignment as well?

Reply:
I've been thinking about defining `Image` as:

```python
@attr.define
class Image(Item):
    url: Union[str, _Url] = attrs.field(converter=url_to_str)
```

However, this part of the code ([reference](https://github.com/zytedata/zyte-common-items/blob/main/zyte_common_items/base.py#L89)) doesn't allow `Union`s in the field types and would raise a `ValueError`. This constraint is needed to be deterministic when converting using `.from_dict()` and `.from_list()`.

One approach we could take to is to have the `web-poet` URL classes inherit from `str` but this would affect our future plans of using `yarl` or `w3lib` in the future. 

Another alternative is to update our [code ](https://github.com/zytedata/zyte-common-items/blob/main/zyte_common_items/base.py#L89) to accept `Union`s but follows a priority for conversion (e.g. `Union[str, int]` would try to auto-convert the value to `str` first and then `int`). 

Perhaps there are better ways to do this so I'll continue to explore them. Kindly let me know if you also think of one. :) 

Nonetheless, despite this **mypy** error, the auto-conversion of `web-poet`'s URL classes into `str` is still supported as seen in the [conversion tests](https://github.com/zytedata/zyte-common-items/pull/11/files#diff-1de3b8c60439d23d5e65480ab44d0329f42a1312a85497580aaa3fd5e9ac750eR50-R59).

Reply:
I was hoping that mypy + attrs are smart enough to look into convertors: url attribute must be str (Union would be wrong, it must be str), but an assignment of _Url should work, because it goes through convertor, which returns str. So, I was expecting type validation to use annotations of url_to_str arguments, not annotation of the url attribute.

If that's not working, shouldn't we mark the test as xfail?

Reply:
After looking into this further, it would seem that it's an upstream issue from mypy itself. Marked these specific test cases as xfail and added links to the relevant GH issues in the code comments. https://github.com/zytedata/zyte-common-items/pull/11/commits/635b30daa2106e5448dd8c4ff9a41e9e009960b5